linkerd/stack/src/annotate_error.rs

```diff
@@ -0,0 +1,227 @@
+use crate::{layer, ExtractParam, NewService, Service};
+use std::{
+    error::Error,
+    fmt,
+    future::Future,
+    marker::PhantomData,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+/// A [`NewService`] that extracts an [`ErrorContext`] implementation from its target,
+/// and produces an [`AnnotateError`] middleware which wraps an inner [`Service`]'s
+/// error type with that [`ErrorContext`].
+pub struct NewAnnotateError<N, C, X = ()> {
```

Comment:
I have a slight bias to keeping `N` as the last param... In our stack the inner type can be very very large, so i try to keep the contextual types first so they're easier to spot in compile errors... e.g. `NewAnnotateError<MyErrorContext, (), Some<Very<Deep<Type<Signature<...>>>>>>`

That may be worth

Reply:
hmm, good point --- i put the `ExtractParam` type last so that it could default to `()`, but I don't think the default is actually that important. I'll change it.